tests/test_minio_repo.py

```diff
@@ -0,0 +1,149 @@
+        self.obj['Body'].read.return_value = b'{}'
+        subscription = Subscription(self.obj, 'some_name', now=datetime.utcnow())
+        assert not subscription.is_valid
+        assert subscription.error == "data missing required key:'c'"
+
+    def test_subscription__when_missing_expiration__should_be_valid(self):
+        self.obj['Body'].read.return_value = b'{"c": "http://callback.com/1"}'
+        subscription = Subscription(self.obj, 'some_name', now=datetime.utcnow())
+        assert subscription.is_valid
+
+
+class SubscriptionsRepoTest(TestCase):
+    def setUp(self):
+        boto3_patch = mock.patch('libtrustbridge.repos.miniorepo.boto3')
+        self.boto3 = boto3_patch.start()
+        self.client = self.boto3.client.return_value
+        self.addCleanup(boto3_patch.stop)
+        self.connection_data = {
+            'access_key': 'awsAccessKeyId',
+            'secret_key': 'awsSecretAccessKey',
+            'host': 'some_host',
+            'port': '1111',
+            'use_ssl': False,
+        }
+
+    def test_subscribe_by_id__should_put_object_into_repo(self):
+        repo = SubscriptionsRepo(connection_data=self.connection_data)
+        id = Id('some_ref')
+
+        repo.subscribe_by_id(id, 'http://callback.url/1')
+        self.client.put_object.assert_called_once()
+        args, kwargs = self.client.put_object.call_args
+
+        assert kwargs['Bucket'] == 'subscriptions'
+        assert kwargs['ContentLength'] == 41
+        assert kwargs['Key'] == 'some_ref'
+        assert kwargs['Body'].read() == b'{"c": "http://callback.url/1", "e": null}'
+
+    def test_subscribe_by_pattern__should_put_object_into_repo(self):
+        repo = SubscriptionsRepo(connection_data=self.connection_data)
+        pattern = Pattern('aaa.bbb.ccc')
+
+        repo.subscribe_by_pattern(pattern, 'http://callback.url/1')
+        self.client.put_object.assert_called_once()
+        args, kwargs = self.client.put_object.call_args
+
+        assert kwargs['Bucket'] == 'subscriptions'
+        assert kwargs['ContentLength'] == 41
+        assert kwargs['Key'] == 'AAA/BBB/CCC/ff0d1111f6636c354cf92c7137f1b5e6'
+        assert kwargs['Body'].read() == b'{"c": "http://callback.url/1", "e": null}'
+
+    def test_get_subscription_by_id__should_return_subscriptions(self):
+        repo = SubscriptionsRepo(connection_data=self.connection_data)
+        self.client.list_objects.return_value = {
+            'Contents': [{'Key': 'some_ref'}]
+        }
+        self.client.get_object.return_value = {
+            'Body': BytesIO(b'{"c": "http://callback.url/1", "e": null}'),
+            'Bucket': 'subscriptions',
+            'ContentLength': 39,
+            'Key': 'some_ref',
+        }
+
+        subscriptions = repo.get_subscriptions_by_id(Id('some_ref'))
+
+        assert list(subscriptions)[0].callback_url == 'http://callback.url/1'
+        self.client.list_objects.assert_called_once_with(Bucket='subscriptions', Prefix='some_ref')
```

Comment:
@adamkerz I'm not sure if we need to search subscriptions by prefix when getting by id, what do you think?

Reply:
I think we do. Subscriptions will be (well, are currently) stored as json files in a directory (ignoring s3 terminology). The name of the file is the hash of the callback url. The name of the directory (or heirarchy of directories, if we split an ID into 5 char chunks) is the ID. Multiple subscriptions can exist for a single ID:

`/abcd-1234-fedc-9876/1b32b3bdfe3f`:
```
url="http://blah/tellMe"
timeout=199912394832948
```

`/abcd-1234-fedc-9876/12d3871237`:
```
url="http://foo/i_wanna_know"
timeout=154439948334
```

So we can't do `get_object` (because we don't know the hash of each subscriber) but need to do `list_objects` to get all subscribers.

Reply:
also please note I've had some problems when list_objects and get_object work differently if you pass or not pass trailing slash or starting slash for Minio and the real S3 (something like Minio considering "/a/b/c" and "a/b/c" the same while S3 is not). Just to mention